controller/events.go

```diff
@@ -0,0 +1,245 @@
+package main
+
+import (
+	"container/list"
+	"errors"
+	"fmt"
+	"strconv"
+	"strings"
+	"sync"
+
+	"github.com/flynn/flynn/Godeps/_workspace/src/gopkg.in/inconshreveable/log15.v2"
+	ct "github.com/flynn/flynn/controller/types"
+)
+
+// EventQueue is a FIFO queue for storing app events (used by EventSubscriber
+// so that slow receivers don't block the main EventListener loop).
+type EventQueue struct {
+	events *list.List
+	cond   *sync.Cond
+	closed bool
+}
+
+func newEventQueue() *EventQueue {
+	return &EventQueue{
+		events: list.New(),
+		cond:   sync.NewCond(&sync.Mutex{}),
+	}
+}
+
+// Push pushes an event onto the tail of the queue.
+func (e *EventQueue) Push(event *ct.AppEvent) {
+	e.cond.L.Lock()
+	e.events.PushBack(event)
+	e.cond.L.Unlock()
+	e.cond.Signal()
+}
+
+// Pop removes and returns an event from the head of the queue, returning
+// nil if the queue is closed.
+func (e *EventQueue) Pop() *ct.AppEvent {
+	e.cond.L.Lock()
+	defer e.cond.L.Unlock()
+	if e.closed {
+		return nil
+	}
+	a := e.events.Front()
+	for a == nil {
+		e.cond.Wait()
+		if e.closed {
+			return nil
+		}
+		a = e.events.Front()
+	}
+	e.events.Remove(a)
+	return a.Value.(*ct.AppEvent)
+}
+
+// Close signals Pop to return nil (which should be treated as an end-of-queue marker).
+func (e *EventQueue) Close() {
+	e.cond.L.Lock()
+	e.closed = true
+	e.cond.L.Unlock()
+	e.cond.Signal()
+}
+
+// EventSubscriber receives app events from the EventListener loop and maintains
+// it's own loop to forward those events to the Events channel.
+type EventSubscriber struct {
+	Events chan *ct.AppEvent
+	Err    error
+
+	l          *EventListener
+	queue      *EventQueue
+	appID      string
+	objectType string
+	objectID   string
+}
+
+// Notify filters the event based on it's type and objectID and then pushes
+// it to the event queue.
+func (e *EventSubscriber) Notify(event *ct.AppEvent) {
+	if e.objectType != "" && e.objectType != string(event.ObjectType) {
+		return
+	}
+	if e.objectID != "" && e.objectID != event.ObjectID {
+		return
+	}
+	e.queue.Push(event)
```

Comment:
After thinking about this a bit, it seems like a buffered channel is actually better, as it would reduce the amount of code used by a significant amount (use a `select` to send with a `default` case that closes the client connection with a "slow reader" error)